File: solution/python3/142.linked-list-cycle-ii.py
#
# @lc app=leetcode id=142 lang=python3
#
# [142] Linked List Cycle II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        walker = head
        runner = head
        while True:
            if not runner or not runner.next or not runner.next.next:
                return None

            walker = walker.next
            runner = runner.next.next
            if walker == runner:
                break

        while True:
            if walker == head:
                return head
            walker = walker.next
            head = head.next

# @lc code=end

File: solution/java8/142.linked-list-cycle-ii.java
/*
 * @lc app=leetcode id=142 lang=java
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null) {
            return null;
        }

        ListNode walker = head;
        ListNode runner = head;
        while (true) {
            if (walker.next != null) {
                walker = walker.next;
            } else {
                return null;
            }

            if (runner.next != null && runner.next.next != null) {
                runner = runner.next.next;
            } else {
                return null;
            }

            if (walker.equals(runner)) {
                break;
            }
        }

        while (true) {
            if (head.equals(walker)) {
                return head;
            }
            walker = walker.next;
            head = head.next;
        }
    }
}
// @lc code=end

